noxfile.py

```diff
@@ -195,27 +241,55 @@ def build_release(session):
 
     session.log("# Ensure no files in dist/")
     if release.have_files_in_folder("dist"):
-        session.error("There are files in dist/. Remove them and try again")
+        session.error(
+            "There are files in dist/. Remove them and try again. "
+            "You can use `git clean -fxdi -- dist` command to do this"
+        )
 
     session.log("# Install dependencies")
     session.install("setuptools", "wheel", "twine")
 
-    session.log("# Checkout the tag")
-    session.run("git", "checkout", version, external=True, silent=True)
+    with mk_tmp_git_checkout(session, version) as build_dir_path:
+        with workdir(session, build_dir_path):
+            build_dists(session)
+
+        tmp_dist_dir = build_dir_path / 'dist'
+        session.log(f"# Copying dists from tmp {tmp_dist_dir}")
+        shutil.copytree(tmp_dist_dir, 'dist')
 
+
+def build_dists(session):
     session.log("# Cleanup build/ before building the wheel")
     if release.have_files_in_folder("build"):
         shutil.rmtree("build")
 
+    session.log(
+        "# Check if there's any Git-untracked files before building the wheel",
+    )
+    has_git_untracked_files = any(
+        bool(l) for l in
+        # session.run doesn't seem to return any output
+        subprocess.check_output(
+            (
+                "git", "ls-files", "--ignored", "--exclude-standard",
+                "--others", "--", ".",
+            ),
+            text=True,
+        ).split('\n')
+        if not l.startswith('.nox/build-release/')  # exclude nox env file
+    )
```

Comment:
This `True` bit is incorrect — `bool(fname)` is supposed to filter out empty strings, also I was trying to avoid converting to a list in-memory so that `any()` would stop iterating early. I'll use @uranusjr's advice here.

As for `filter()` — I think it's a lot of people don't like it when a functional style is combined with comprehensions. I'll stick with comprehensions I think.

P.S. What does `None` do? Is it the same as `bool` in this context?

Reply:
Yea, filter(None, ...) filters out falsy items.

Reply:
Anyway... This is not really intuitive, somebody in the future may decide to "refactor" it and easily get it wrong :)

Reply:
`filter(None, ...)` isn't necessary, since `any` checks for truthiness, not boolean `True`, so it ends up being

```
cmd = ["git", "ls-files", "--ignored", "--exclude-standard", "--others", "--", "."]
all_untracked_files = subprocess.check_output(cmd, text=True).splitlines()
abort_due_to_untracked_files = any(
    f for f in all_untracked_files if not f.startswith(".nox/build-release)
)
```

Reply:
@chrahunt nope, there's a corner case: you can an empty string if there's no rogue files. I don't remember the details but the behavior was surprising so I'm doing it explicitly now.

Reply:
But `any([""])` is false, which is the expected behavior, right? If what I put wouldn't work I'd rather see a comment describing why because that's the first thing that I'd do seeing this.

Reply:
Well, I don't remember what I hit exactly but I guess it doesn't matter now that it's rewritten a few times after that already :)